### ISSUE TITLE
Malformed XML output after gapseq fill

### ISSUE THREAD

**Issue by jonasoh** [jotech/gapseq#149]
Hi,
I'm getting malformed XML output from one of my models. I have tested it on both MacOS and in a Ubuntu 22.04 VM (both on ARM). Trying to gapfill using `dat/media/ALLmed.csv`, the error message looks something like this on MacOS:
```
Patching file MAG5.xml ... done
Warning message:
In readLines(paste0(out.id, ".xml")) :
  line 25578 appears to contain an embedded nul
```

And like this on Linux:
```
Model file already exists and will be overwritten! 
*** buffer overflow detected ***: terminated
/home/j/gapseq/gapseq: line 145: 34485 Aborted                 (core dumped) Rscript $dir/src/gf.suite.R "$@"
```

In the XML file, the error is apparent on line 25578:
```
      </species>
      <species metaid="M_cpd15504_c0" id="M_cpd15504_c0" name="CoaBC_6_379
        <annotation>
```

Trying to gapfill with `dat/media/MM_anaerobic_Acetate.csv` causes the R process to hang at 100% CPU utilization after displaying `1. Initial gapfilling: Make model grow on given media using all reactions`. 

Gapfilling with `dat/media/LBmed.csv` again gives a malformed XML file, this time on line 25561:
```
      </species>
      <species metaid="M_cpd15509_c0" id="M_cpd15509_c0" name="CoaBC_6_379
        <annotation>
```

Any clue what is going on here? Some kind of memory corruption in sybilSBML/libSBML, or something in gapseq? Here are the relevant files, including the RDS model from LBmed filling: [MAG5-files.zip](https://github.com/jotech/gapseq/files/9774615/MAG5-files.zip)

**Comment by jonasoh** [comment]
Adding to this, I get the same results after reconstructing the model using a protein FASTA of the same organism. It happens with most media, but not all.

**Comment by Waschina** [comment]
I will look into this. Does the final gap-filled XML also have the formatting error?

**Comment by jonasoh** [comment]
Yes, it does. I'm stuck on what is actually going on here. It seems like the process is crashing during the "patching" phase of sbml-write.R. If I manually run the code in RStudio, it doesn't crash. The resulting files are then identical, except for the truncated line:
```
% diff MAG5.xml MAG5_manual.xml
25526c25526
<       <species metaid="M_cpd15510_c0" id="M_cpd15510_c0" name="CoaBC_6_379
---
>       <species metaid="M_cpd15510_c0" id="M_cpd15510_c0" name="two linked disacharide pentapeptide and tetrapeptide murein units (uncrosslinked, middle of chain)-c0" compartment="c0" hasOnlySubstanceUnits="false" boundaryCondition="false" constant="false" fbc:charge="-4" fbc:chemicalFormula="C77H119N15O41">
```

The truncated line always ends with the string CoaBC_6_379:
```
% grep -C1 CoaBC *.xml
MAG5.xml-      </species>
MAG5.xml:      <species metaid="M_cpd03494_c0" id="M_cpd03494_c0" name="CoaBC_6_379
MAG5.xml-        <annotation>
--
MAG5_ALLmed.xml-      </species>
MAG5_ALLmed.xml:      <species metaid="M_cpd15504_c0" id="M_cpd15504_c0" name="CoaBC_6_379
MAG5_ALLmed.xml-        <annotation>
--
MAG5_LBmed.xml-      </species>
MAG5_LBmed.xml:      <species metaid="M_cpd15509_c0" id="M_cpd15509_c0" name="CoaBC_6_379
MAG5_LBmed.xml-        <annotation>
```

**Comment by jonasoh** [comment]
No, the corruption occurs during writing of the unpatched file, which corrupts the file at a certain position, causing truncation upon reading the file for patching, should have caught that earlier. This is what it looks like after  `sybilSBML::writeSBML`, running the `write_gapseq_sbml` function from R in the terminal:
<img width="877" alt="bild" src="https://user-images.githubusercontent.com/6480370/196482770-0bb1c1de-ba1c-472e-a3eb-7a77c36ed8b7.png">

And this is what it looks like if I do the exact same thing in RStudio (using the same R installation):
<img width="877" alt="bild" src="https://user-images.githubusercontent.com/6480370/196483431-112e4964-d95e-4e37-982f-64b7b92cbb17.png">

I guess the problem is thus most likely in sybil-/libSBML, and that it is triggered by something in the environment which differs between the terminal and RStudio.

**Comment by Waschina** [comment]
That's interesting... Especially that it always happens with that specific metabolite.

When to run the code in RStudio, R and Rstudio are crashing. And I am wondering, where this CoaBC_6_379 is coming from.

**Comment by jonasoh** [comment]
Which metabolite is affected can actually vary depending on the gapfilling medium, and also clusters around certain points, e.g., metabolites cpd15504, cpd15506, cpd15509 are affected in different media. Very strange indeed... 

I now updated my installation of libsbml from 5.19.0 to 5.19.7, and this also changes the affected metabolite. Thus it's pretty clear that the problem lies in libsbml I think.

**Comment by Waschina** [comment]
I noticed in your model, that the gene names could be the reason for the issue. Three examples of gene names:
```
> mod@genes[25]
[[1]]
[1] "MAG5_FNLNEH_05175_Alpha-hydroxy-acid_oxidizing_protein_1:334"

> mod@genes[24]
[[1]]
[1] "MAG5_FNLNEH_10140_Adenosylhomocysteinase_2:410"

> mod@genes[23]
[[1]]
[1] "MAG5_FNLNEH_10965_NAD" "+"                     "_synthase_13:240"     
```

Apparently, special characters in the sequence headers of the input fasta can cause errors in the gapseq workflow.

Ideally, the fasta headers of the sequences should look have the form `>UNIQUEID Gene description` . So in Your case for instance: `MAG5_FNLNEH_10965 NAD+_synthase` instead of  `MAG5_FNLNEH_10965_NAD+_synthase` (Note the space).

But also in gapseq, we should make sure that special characters in the UNIQUEID-Part do not cause errors like the one above (Nr. 23)... I'll try to fix that in the next days.

**Comment by jonasoh** [comment]
Interesting... The gene name for this particular one in the FASTA file actually looks like this:
```
>MAG5_FNLNEH_10965_NAD(+)_synthase
```

But then I would assume something like this would fix the issue, right?
```
for (i in 1:length(mod@genes)) { 
    mod@genes[[i]] <- paste0(mod@genes[[i]], sep='', collapse='')
}

> mod@genes[23]
[[1]]
[1] "MAG5_FNLNEH_10965_NAD+_synthase_13:240"
```
But this actually introduces further errors in the file:
<img width="877" alt="bild" src="https://user-images.githubusercontent.com/6480370/196878626-38ed7aa6-86df-4f23-a32c-667467a66e61.png">

And more troubling, there are silent errors, i.e., correctly-formed XML with incorrect data, introduced into the file (those might have been present earlier as well):
<img width="877" alt="bild" src="https://user-images.githubusercontent.com/6480370/196879033-ce0219b1-ae96-42f9-aef3-b353f7debab2.png">

**Comment by Waschina** [comment]
The gene names should be fixed already in the draft script. I guess the brackets and possibly other characters cause errors in string handling when gapseq assigns gene names based on sequence headers.

The gene identifier handling in gapseq is also explained and discussed in issue #71 .

Changing the Object `mod@genes` after the model is already reconstructed (draft) unfortunately is not a solution as this breaks the consistency with the other gene-related entries in the S4-object `mod`, i.e.
```R
mod@gprRules
mod@gpr
mod@rxnGeneMat
mod@allGenes
``` 

I'll try to fix the wrong string handling in gapseq today. Perhaps you could try to rerun the draft script for your MAG5 again? This gene naming could also be an independent issue from the wrong xml-formatting, but somehow I think it is related. Let's see

**Comment by jonasoh** [comment]
Ah, I see... great work! Your suggested fix sounds good to me.

I misunderstood you previously regarding the space issue. That seems like the proper way to format the file for sure.

**Comment by jonasoh** [comment]
I reran gapseq on the properly formatted protein FASTA and it does indeed work as expected. Thanks!

**Comment by Waschina** [comment]
Perfect! I'm glad we solved this. With the last commit, also long titles with special characters should not cause errors anymore.

**Comment by jonasoh** [comment]
Excellent. I am still puzzled by the fact that the script ran fine in RStudio, but maybe that will simply remain a mystery.
